### DISCUSSION TITLE
Is there an unmount API and lifecycle method?

### DISCUSSION THREAD

**Discussion by zidell** [vanjs-org/van#78] [Q&A] [answered]
Hi, this is a very cool project!

I was wondering if there is a unmount API that corresponds to `van.add(...)`.
Also, are there any lifecycle methods like created, mounted, willunmount, and so on?

Any answers would be appreciated.

**Comment by Tao-VanJS** [comment, marked as answer]
Hi @zidell,

Thanks for reaching out!

In VanJS, lifecycle event hooks can be easily implemented with the help of a lifecycle state and `van.derive`. For example:

```js
const stage = van.state("init")

van.add(document.body, () => {
  if (stage.val === "init") return div()
  if (stage.val === "mounted") return div(
    <component content>
  )
  if (stage.val === "unmounted") return null
})

van.derive(() => {
  if (stage.val === "mounted") {
    <willmount hook>
  } else if (stage.val === "unmounted") {
    <willunmount hook>
  }
})
```

The code can be slightly adjusted if you want hooks in more lifecycle stages.

As a minimalist framework, VanJS won't include features that can be easily achieved with the combination of other features.

Hope it helps :-)

**Reply by efpage** [reply]
Lifecycle methods maybe necessary to prevent state updates before the page build is finished or before a component is fully ready to use. In many cases it would be sufficient to be able to intercept the state propagation until an operation is finished. This could a very lightweight solution.

**Reply by zidell** [reply]
Thanks for the answer and detailed examples. I can think of a lot of ways to apply it that way.

What I wanted was a slightly different case. My webpage dynamically replaces content via pjax. And vanjs can only work on this particular page. If the user navigates to a different page, the vanjs itself might disappear and something else might be displayed. 

The code above assumes that the van component is still there, so it looks like it's just changing the display.

Is there a way to remove everything associated with the mounted component(including memory, events, etc) as if it never existed?

If there are no plans to add a lifecycle API, can you share some ideas on implementing something like this?

**Reply by Tao-VanJS** [reply]
Hi @zidell,

Thanks for explaining your use case!

In function-based binding of VanJS, if the function returns `null` or `undefined`, it means the bound DOM node will be removed permanently. You can refer to https://vanjs.org/tutorial#removing-a-dom-node for the details. Again, you can use `van.derive` to register event handlers.

Let me know if it helps:-)

**Reply by zidell** [reply]
I applied the code you gave me and did the following.

```js
// I arbitrarily extended the function for mounting.
van._mount = function (target, component) {
	if (!!target) {
		van._mount.__ = van.state(true);
		van.add(target, () => {
			if (!van._mount.__.val) {
				return null;
			}
			return component();
		});
	} else {
		van._mount.__.val = false;
	}
};

// Make a component with a function that depends on the lifecycle state as follows.
var ExampleComponent = () => {
	van.derive(() => {
		if (van._mount.__.val) {
			console.log('mounted');
		} else {
			console.log('will unmount');
		}
	});
	return van.tags.div('this is an UserComponent');
};

// Mount via extended function
van._mount(document.querySelector('#app'), ExampleComponent);

// When we need to control this component externally, we call it like this
// van._mount(null);
```

This works fine for my intentions.
https://codepen.io/zidell/pen/YzRmZZy

Thank you very much.

**Reply by Tao-VanJS** [reply]
You're welcome 😀

I'm glad that the code can work for you.

**UPDATE: with VanJS 1.1.0, we don't have the restriction described below any more.**

~~There is one more thing: it would be better to move the `van.derive` call out of `ExampleComponent`. Essentially, `ExampleComponent` is a binding function for DOM nodes. And if inside the binding function, the derivation function specified in `van.derive` refers to a state defined outside the binding function, this might lead to memory leaks if the binding function `ExampleComponent` is called many times.~~

~~You can refer to this section: https://vanjs.org/advanced#derived-states-and-side-effects-are-not-subject-to-gc as a detailed explanation.~~

**Reply by Tao-VanJS** [reply]
I'm considering to remove this restriction in an upcoming release.

**Reply by Tao-VanJS** [reply]
With VanJS [`1.1.0` release](https://github.com/vanjs-org/van/discussions/81), we don't have this restriction anymore.
